File: examples/boundary_conditions/predefined_BC/plate.cc
/**
 * @file   plate.cc
 *
 * @author Dana Christen <dana.christen@epfl.ch>
 * @author Aurelia Isabel Cuba Ramos <aurelia.cubaramos@epfl.ch>
 * @author Nicolas Richart <nicolas.richart@epfl.ch>
 *
 * @date creation: Wed Dec 16 2015
 *
 * @brief  boundary condition example
 *
 * @section LICENSE
 *
 * Copyright (©) 2015 EPFL (Ecole Polytechnique Fédérale de Lausanne) Laboratory
 * (LSMS - Laboratoire de Simulation en Mécanique des Solides)
 *
 * Akantu is free  software: you can redistribute it and/or  modify it under the
 * terms  of the  GNU Lesser  General Public  License as  published by  the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Akantu is  distributed in the  hope that it  will be useful, but  WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A  PARTICULAR PURPOSE. See  the GNU  Lesser General  Public License  for more
 * details.
 *
 * You should  have received  a copy  of the GNU  Lesser General  Public License
 * along with Akantu. If not, see <http://www.gnu.org/licenses/>.
 *
 */

/* -------------------------------------------------------------------------- */
#include "solid_mechanics_model.hh"
/* -------------------------------------------------------------------------- */
#include <iostream>
/* -------------------------------------------------------------------------- */

using namespace akantu;

int main(int argc, char *argv[]) {
  initialize("material.dat", argc, argv);

  //Do I need to define the element type?
  UInt spatial_dimension = 2;
  UInt max_steps = 3000;

  Mesh mesh(spatial_dimension);
  mesh.read("square.msh");
  mesh.createGroupsFromMeshData<std::string>("physical_names");

  SolidMechanicsModel model(mesh);

  /// model initialization
  model.initFull();

  std::cout << model.getMaterial(0) << std::endl;
  model.assembleMassLumped();

  /// Dirichlet boundary conditions
  model.applyBC(BC::Dirichlet::FixedValue(0.0, _x), "Fixed_x");
  model.applyBC(BC::Dirichlet::FixedValue(0.0, _y), "Fixed_y");

  model.updateResidual();

  model.setBaseName("plate");
  model.addDumpFieldVector("displacement");
  model.addDumpField("blocked_dofs");
  model.addDumpField("force"       );
  model.addDumpField("residual"    );
  model.dump();

  Real time_step = model.getStableTimeStep() * .8;
  model.setTimeStep(time_step);

  Vector<Real> traction(2);
  for(UInt s = 1; s <= max_steps; ++s) {
    if(s % 100 == 0)
      std::cout << "passing step " << s << "/" << max_steps << std::endl;

    // Neumann boundary condition
    traction(0) = s*0.1;
    traction(1) = 0.0;
    model.applyBC(BC::Neumann::FromTraction(traction), "Traction");

    model.explicitPred();
    model.updateResidual();
    model.updateAcceleration();
    model.explicitCorr();

    if(s % 10 == 0)
      model.dump();
  }

  finalize();
  return EXIT_SUCCESS;
}

File: examples/python/plate-hole/plate.py
#!/usr/bin/env python3
try:
    from mpi4py import MPI
    comm = MPI.COMM_WORLD
    prank = comm.Get_rank()
except ImportError:
    prank = 0

import akantu as aka
import numpy as np


# -----------------------------------------------------------------------------
def solve(material_file, mesh_file, traction):
    aka.parseInput(material_file)
    spatial_dimension = 2

    # -------------------------------------------------------------------------
    # Initialization
    # -------------------------------------------------------------------------
    mesh = aka.Mesh(spatial_dimension)
    if prank == 0:
        mesh.read(mesh_file)

    mesh.distribute()

    model = aka.SolidMechanicsModel(mesh)
    model.initFull(_analysis_method=aka._static)

    model.setBaseName("plate")
    model.addDumpFieldVector("displacement")
    model.addDumpFieldVector("external_force")
    model.addDumpField("strain")
    model.addDumpField("stress")
    model.addDumpField("blocked_dofs")

    # -------------------------------------------------------------------------
    # Boundary conditions
    # -------------------------------------------------------------------------
    model.applyBC(aka.FixedValue(0.0, aka._x), "XBlocked")
    model.applyBC(aka.FixedValue(0.0, aka._y), "YBlocked")

    model.getExternalForce()[:] = 0
    trac = np.zeros(spatial_dimension)
    trac[1] = traction
    model.applyBC(aka.FromTraction(trac), "Traction")

    solver = model.getNonLinearSolver()
    solver.set("max_iterations", 2)
    solver.set("threshold", 1e-10)
    solver.set("convergence_type", aka.SolveConvergenceCriteria._residual)

    print("Solve for traction ", traction)
    model.solveStep()

    model.dump()


# -----------------------------------------------------------------------------
# main
# -----------------------------------------------------------------------------
def main():
    mesh_file = 'plate.msh'
    material_file = 'material.dat'
    traction = 1.

    solve(material_file, mesh_file, traction)


# -----------------------------------------------------------------------------
if __name__ == "__main__":
    main()
